Sprint 7 solution in file manager

MadokaKaname42 opened the pull request.

MadokaKaname42 committed:
Во все сущности добавлены поля типа TaskType;

MadokaKaname42 committed:
Создано перечисление TaskType

MadokaKaname42 committed:
Изменены модификаторы доступа, для полей

MadokaKaname42 committed:
Создан класс, FileBackedTaskManager, для чтения и записи в файл

MadokaKaname42 committed:
Тесты для класса FileBackedManager

MadokaKaname42 committed:
CodeStyle

vlolad commented on src/model/Task.java:
```diff
@@ -9,6 +10,7 @@ public class Task {
     private String description;
     private Status status;
     private int id = 0;
+    private TaskType type = TaskType.TASK;
```
Для типа задачи лучше обойтись вообще без поля - его тяжело переопределять в наследниках. Лучше оставь метод `getType()`, который будет сразу возвращать тип задачи текущего класса:
```
    public TaskType getType() {
        return TaskType.TASK;
    }
```

В класса-наследниках будет достаточно переопределить метод, и доп. поле не потребуется

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,178 @@
+            e.printStackTrace();
+        }
+
+        return manager;
```
Надо ещё перезаписывать значение `counterId`, сейчас если в новом менеджере создать задачу, то будет коллизия по id
Можно сделать поле counterId в менеджере protected, а здесь - отследить максимальный id у загружаемой задачи с помощью доп. поля и его присваивать после парсинга всех строк из файла

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    public void save() {
+        try (FileWriter writer = new FileWriter("tasks.csv")) {
```
Название файла (и первую строку в файле с форматом CSV) лучше вынести в константу

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    public static Task fromString(String value) {
```
Методы для конвертации задач в/из строки лучше вынести в отдельный класс-конвертер (методы можно сделать публичными статическими). Преобразование задач не прямая задача менеджера, плюс форматы могут быть разные, удобнее будет править их в отдельном классе

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    private String toString(Task task) {
```
Название метода надо поменять, сейчас хоть и есть параметр, но нейминг дублирует стандартный метод класса Object, это плохая практика

Можно назвать taskToString

vlolad reviewed (changes requested):
Привет!

Работает хорошо. Один баг с id есть, в остальном принято. По парсингу и сериализации в строку замечаний нет

По коду оставил комментарии насчёт рефакторинга, но есть одно общее замечание - надо в классе Managers изменить метод `getDefault()` - он теперь должен возвращать новую реализацию менеджера

Хороших выходных!

MadokaKaname42 committed:
Удалено поле TaskType;
Изменены методы getType;

MadokaKaname42 committed:
Метод toString() изменен под формат записи в файл

MadokaKaname42 committed:
Создан класс-конвертер

MadokaKaname42 committed:
Изменен модификатор доступа для поля counterId

MadokaKaname42 committed:
Метод FromString перенесен в класс StringConverter;
Изменения для метода loadFromString для избежания конверсии id;
Метод toString() теперь используется из реализации в классах Task/Subtask/Epic

vlolad commented on src/model/Epic.java:
```diff
@@ -1,12 +1,24 @@
+        super(id, name, description, status);
+    }
+
+    public TaskType getType() {
```
`@Override` обязательно для переопределенных методов

vlolad commented on src/model/SubTask.java:
```diff
@@ -23,8 +30,19 @@ public void setEpicId(int epicId) {
         this.epicId = epicId;
     }
 
+    public TaskType getType() {
```
`@Override` обязательно для переопределенных методов

vlolad commented on src/tools/StringConverter.java:
```diff
@@ -0,0 +1,30 @@
+import model.SubTask;
+import model.Task;
+
+public class StringConverter {
```
Надо сюда же вынести метод для преобразования задачи в строку. toString из метода Task и наследников использовать - плохая идея, по всё тем же причинам. Форматы могут быть разные + при изменении количества строк придётся исправлять три класса вместо одного

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,141 @@
+import java.io.FileWriter;
+import java.io.IOException;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
Пропущено замечание из общего комментария, так что дублирую сюда)

Есть одно общее замечание - надо в классе Managers изменить метод `getDefault()` - он теперь должен возвращать новую реализацию менеджера

vlolad reviewed (changes requested):
Привет!

Пропущено одно замечание по поводу класса Managers, плюс надо доработать конвертер задач. Остальное по мелочи, override-аннотации добавить

Постараюсь принять ASAP с доработками, так как правок нужно немного
Хорошего дня!

MadokaKaname42 committed:
Добавлены аннотации Override переопределенным методам getType()

MadokaKaname42 committed:
Реализован метод toString()

MadokaKaname42 committed:
Метод getDefault теперь возвращает реализацию FileBackedTaskManager()

MadokaKaname42 committed:
метод save() теперь использует реализацию toString() из класса StringConverter

MadokaKaname42 committed:
codeStyle

MadokaKaname42 committed:
remove Unused import

vlolad reviewed (approved):
Привет!

Всё, теперь принято
